src/result_processor.c

```diff
@@ -913,18 +920,85 @@ static ResultProcessor *RPSafeLoader_New(RedisSearchCtx *sctx, RLookup *lk, cons
 
 /*********************************************************************************/
 
-#define DEFAULT_BUFFER_BLOCK_SIZE 1024
-
 ResultProcessor *RPLoader_New(AREQ *r, RLookup *lk, const RLookupKey **keys, size_t nkeys) {
+  r->stateflags |= QEXEC_S_HAS_LOAD;
   if (r->reqflags & QEXEC_F_RUN_IN_BACKGROUND) {
     // Assumes that Redis is *NOT* locked while executing the loader
-    return RPSafeLoader_New(r->sctx, lk, keys, nkeys, DEFAULT_BUFFER_BLOCK_SIZE);
+    return RPSafeLoader_New(r->sctx, lk, keys, nkeys);
   } else {
     // Assumes that Redis *IS* locked while executing the loader
     return RPPlainLoader_New(r->sctx, lk, keys, nkeys);
   }
 }
 
+// Consumes the input loader and returns a new safe loader that wraps it.
+static ResultProcessor *RPSafeLoader_New_FromPlainLoader(RPLoader *loader) {
+  RPSafeLoader *sl = rm_new(RPSafeLoader);
+
+  // Copy the loader, move ownership of the keys
+  sl->base_loader = *loader;
+  rm_free(loader);
+
+  // Reset the loader's buffer and state
+  sl->BufferBlocks = NULL;
+  sl->buffer_results_count = 0;
+  sl->curr_result_index = 0;
+
+  sl->last_buffered_rc = RS_RESULT_OK;
+
+  sl->base_loader.base.Next = rpSafeLoaderNext_Accumulate;
+  sl->base_loader.base.Free = rpSafeLoaderFree;
+  sl->base_loader.base.type = RP_SAFE_LOADER;
+  return &sl->base_loader.base;
+}
+
+void SetLoadersForBG(AREQ *r) {
+  if (r->qiter.endProc->type == RP_SAFE_LOADER) {
```

Comment:
1. Is there a way to unify this with the while loop?
2. Is there a reason why the order of conditions is inverted inside the while?
3. why does only the RP_SAFE_LOADER case worth assert?

Reply:
1. I'll change it a bit, but we must update `r->qiter.endProc` if the last RP was a loader, so there will always be some handling to do
2. No. Not relevant after fixing 1.
3. Because this is an edge case. We usually don't expect to see a safe loader when we re-prepare the pipeline for BG execution, which means the pipeline was converted more than once. The assertion shows that although this is a safe loader, it is set as a regular loader. We can remove it now that we know the tests are passing. WDYT?

Reply:
1. great
3. i support having an assertion check there. I meant why isn't there **also** an assert on the RP_LOADER case

Reply:
Oh. So because the simple loader is not expected to have any special cases like having its `Next` function set to another RP function